# ML-02 — Research Question and Provisional Lane
[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Divyanshu10045/flyrank-assignments/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This notebook frames my capstone research question for the Applied Search Intelligence track. The lane below is *provisional* — I can confirm or change it until the end of Week 4. The sections follow the assignment card in order: lane → the decision/action/cost → a quick checked look at the starter data → the limits of what I can claim → self-check. Everything here is framing and observed counts — no modeling yet.

## 1. My lane (or freestyle) and why

Going with **Lane 2 — Refresh / Content Opportunity Scoring.**

The question I want to answer: which pages should someone actually look at first if they're deciding what to refresh, expand, protect, prune, or leave alone? I picked this over the other lanes because it ends in something a person can act on — a reviewer only has so much time each week, and this tells them where to spend it, with a reason attached instead of a black-box score.

I did look at the other lanes first. Lane 1 (signal analysis) felt like it ends in a report, not a to-do list. Lane 3 (clustering) is a useful way to look at the data but doesn't tell anyone what to *do* next. Lane 4 (CTR/engagement) felt like a narrower slice of the same problem I'm already solving in Lane 2. So Lane 2 it is — and if the numbers hold up on the full release, I might frame it more strongly later as "predict the next 30 days from the last 90," which would make for a stronger capstone.

What actually convinced me was running the numbers below — more than half the pages are declining, and most of them still get real traffic. That's not a small list anyone can eyeball.

In [1]:
# Setup once: climb to the repo root so the relative data path works from anywhere this kernel starts.
import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)
else:
    while not os.path.isdir("data/raw") and os.getcwd() != "/":
        os.chdir("..")

import pandas as pd
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

# Why the pool is too big to eyeball: declining pages with real demand, vs. a reviewer's real capacity.
down_with_demand = df[(df["trend_direction"] == "down") & (df["impressions_90d"] >= 100)]
review_capacity = 40  # assumed pages/week a reviewer can actually act on
print(f"declining pages with >= 100 impressions/90d: {len(down_with_demand):,}")
print(f"at ~{review_capacity} pages/week, that pool alone needs ~{len(down_with_demand) / review_capacity:.0f} weeks of review")

declining pages with >= 100 impressions/90d: 13,152
at ~40 pages/week, that pool alone needs ~329 weeks of review


## 2. The question: decision, action, cost of a wrong call

- **Decision it improves:** which page a reviewer opens first. Output is a ranked queue, not a one-off prediction sitting in a spreadsheet.
- **Who acts on it, and how:** a content editor / SEO reviewer, who has a fixed amount of review time each week and works down the list — refresh, expand, protect, prune, or monitor.
- **Cost of getting it wrong:** ranking an unimportant page too high wastes a review session; missing a page that actually matters means it keeps losing traffic unnoticed. These aren't equally bad — missing something important seems worse than wasting time on something minor — so precision near the *top* of the queue matters more than precision further down.
- **Why this isn't just "train a model":** the real deliverable is a ranked, explainable queue a person can act on and trust — the model is one ingredient, not the point. There are a lot of tangled signals here (traffic, position, CTR, content age, competition), and the repo's own starter pipeline shows a trained model beating a hand-written rule on this kind of ranking task (Precision@50 went from 0.240 with the rule to 0.740 with the model — see `outputs/model_report.md`). I'm treating that as a good sign, not proof — I'd want it to hold up again once I'm working with the real warehouse data, with a reason code attached to every recommendation so a reviewer isn't just trusting a number.

In [2]:
# The declining pool is not one uniform mass: it runs from low-demand to excellent-demand pages,
# which is why a flat “declining” flag is not enough — impact must shape the ranking.
import os
while not os.path.isdir("data/raw") and os.getcwd() != "/":
    os.chdir("..")
import pandas as pd
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

tiers = ["no_data", "none", "low", "moderate", "good", "excellent"]
comp = df[df["trend_direction"] == "down"]["impression_tier"].value_counts().reindex(tiers).fillna(0).astype(int)
print(comp.to_string())

impression_tier
no_data         0
none            0
low          5106
moderate     6435
good         4223
excellent     498


## 3. Quick look at the data (2–3 real numbers)

Ran these against the actual starter CSV (30,000 rows, trailing 90-day window):

1. **54.2% of pages are marked as declining** (16,262 of 30,000). That's more than I expected — decline isn't rare here, it's the majority case, so a plain "flag anything declining" approach would flag half the site and help no one prioritize.
2. **13,152 of those declining pages still have real traffic** (≥100 impressions in 90 days) — about 44% of the whole dataset. That's the actual pile someone would have to work through.
3. **That pile alone still pulls in ~80M impressions over 90 days**, median ~1,620 per page. These aren't dead pages nobody sees — they're losing ground while still getting looked at, which is exactly the kind of thing worth catching early.

This is why I think Lane 2 is worth the next several weeks: a sized, real problem (13k+ pages, way more than one person can review by hand) with a real cost to getting the order wrong. These numbers are from the small teaching slice, not the full dataset — I'll re-check this once I'm on the warehouse data in Week 3.

In [3]:
# The 2–3 headline numbers for Section 3, computed in one place.
import os
while not os.path.isdir("data/raw") and os.getcwd() != "/":
    os.chdir("..")
import pandas as pd
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

n = len(df)
down = df["trend_direction"] == "down"
pool = df[down & (df["impressions_90d"] >= 100)]

print(f"rows: {n:,}")
print(f"1) declining pages: {int(down.sum()):,}  = {down.mean() * 100:.1f}% of pages")
print(f"2) declining with >= 100 impressions/90d: {len(pool):,}  = {len(pool) / n * 100:.1f}% of all pages")
print(f"3) pooled 90d impressions of that pool: {int(pool['impressions_90d'].sum()):,}")
print(f"   median 90d impressions per page in that pool: {pool['impressions_90d'].median():.1f}")

rows: 30,000
1) declining pages: 16,262  = 54.2% of pages
2) declining with >= 100 impressions/90d: 13,152  = 43.8% of all pages
3) pooled 90d impressions of that pool: 79,887,612
   median 90d impressions per page in that pool: 1620.5


## 4. Careful words: what I can and can't claim

What I think I can say, based on what I've actually run:
- These are **observed** patterns in this specific snapshot, not universal facts — e.g. "54.2% of pages in this sample are declining," not "54.2% of all web pages decline."
- Any relationship I find is **directional**, not causal — declining pages having demand doesn't tell me *why*.

What I need to be careful not to say:
- That refreshing a page will *cause* it to recover — no experiment here, just observational data, so no cause-and-effect claims.
- Anything about how Google's algorithm actually works — this is search/analytics data, not a look inside a ranking system.
- That `trend_direction` is ground truth — it's a proxy from the current window. My actual target later is closer to "predict the next 30 days from the last 90," a harder problem.
- Nothing here ties back to an actual client, URL, or query — `client_id` / `content_id` are just anonymized grouping IDs.

Basically: describe what the data shows, not what I think it proves.

In [4]:
# Careful-words guard: re-derive the figures and confirm the slice is safe to work with.
import os
while not os.path.isdir("data/raw") and os.getcwd() != "/":
    os.chdir("..")
import pandas as pd
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("rows x cols:", df.shape)
print("distinct clients:", df["client_id"].nunique(), "(pseudonymized ids only)")
risk = [c for c in df.columns if any(k in c.lower() for k in ("url", "query", "name", "domain"))]
print("columns that could expose client data:", risk if risk else "none")

print("\nre-deriving Section 3 figures from this executed cell:")
down = df["trend_direction"] == "down"
pool = df[down & (df["impressions_90d"] >= 100)]
print(f"  declining: {int(down.sum()):,}")
print(f"  declining with >= 100 impressions/90d: {len(pool):,}")
print(f"  pooled 90d impressions: {int(pool['impressions_90d'].sum()):,}")

rows x cols: (30000, 44)
distinct clients: 32 (pseudonymized ids only)
columns that could expose client data: none

re-deriving Section 3 figures from this executed cell:
  declining: 16,262
  declining with >= 100 impressions/90d: 13,152
  pooled 90d impressions: 79,887,612


## 5. Self-check

- Ran this top to bottom, no errors.
- Filled in all four sections with actual reasoning, not just numbers.
- Backed the lane choice with real numbers from the starter data (54.2% declining, 13,152 with real demand, ~80M pooled impressions).
- Double-checked there's nothing client-identifying in here — just anonymized IDs.
- The "40 pages/week" figure in the code above is a guess I used to illustrate the size of the backlog, not based on anything real — worth being upfront about that.